cip36/src/serialization.rs

```diff
@@ -0,0 +1,550 @@
+                            voting_purpose = Some(tmp_voting_purpose);
+                            voting_purpose_encoding = tmp_voting_purpose_encoding;
+                            voting_purpose_key_encoding = Some(key_enc);
+                            orig_deser_order.push(2);
+                        },
+                        (unknown_key, _enc) => return Err(DeserializeFailure::UnknownKey(Key::Uint(unknown_key)).into()),
+                    },
+                    CBORType::Text => return Err(DeserializeFailure::UnknownKey(Key::Str(raw.text()?)).into()),
+                    CBORType::Special => match len {
+                        cbor_event::LenSz::Len(_, _) => return Err(DeserializeFailure::BreakInDefiniteLen.into()),
+                        cbor_event::LenSz::Indefinite => match raw.special()? {
+                            CBORSpecial::Break => break,
+                            _ => return Err(DeserializeFailure::EndingBreakMissing.into()),
+                        },
+                    },
+                    other_type => return Err(DeserializeFailure::UnexpectedKeyType(other_type).into()),
+                }
+                read += 1;
+            }
+            let stake_credential = match stake_credential {
+                Some(x) => x,
+                None => return Err(DeserializeFailure::MandatoryFieldMissing(Key::Uint(1)).into()),
+            };
+            let nonce = match nonce {
+                Some(x) => x,
+                None => return Err(DeserializeFailure::MandatoryFieldMissing(Key::Uint(2)).into()),
+            };
+            if voting_purpose == Some(0) {
+                voting_purpose_default_present = true;
+            }
+            let voting_purpose = voting_purpose.unwrap_or(0);
+            read_len.finish()?;
+            Ok(Self {
+                stake_credential,
+                nonce,
+                voting_purpose,
+                encodings: Some(KeyDeregistrationEncoding {
+                    len_encoding,
+                    orig_deser_order,
+                    stake_credential_key_encoding,
+                    nonce_key_encoding,
+                    nonce_encoding,
+                    voting_purpose_key_encoding,
+                    voting_purpose_encoding,
+                    voting_purpose_default_present,
+                }),
+            })
+        })().map_err(|e| e.annotate("KeyDeregistration"))
+    }
+}
+
+impl Serialize for KeyRegistration {
+    fn serialize<'se, W: Write>(&self, serializer: &'se mut Serializer<W>, force_canonical: bool) -> cbor_event::Result<&'se mut Serializer<W>> {
+        serializer.write_map_sz(self.encodings.as_ref().map(|encs| encs.len_encoding).unwrap_or_default().to_len_sz(4 + if self.voting_purpose != 0 || self.encodings.as_ref().map(|encs| encs.voting_purpose_default_present).unwrap_or(false) { 1 } else { 0 }, force_canonical))?;
+        let (legacy_format, should_include_voting_purpose) = match self.delegation {
```

Comment:
this bit was manually modified. I'm not sure what the desired behavior should even be. Under CDDL `.default` allows you to omit it if it's the default value. The test vectors have it present even as 0, but only in the weighted example. It shouldn't be there if it's canonical CBOR (e.g. for HW wallets I'm guessing?) so we might want an easier way (besides modifying the encodings struct manually). Should it ALWAYS be omitted in legacy examples? The CDDL spec on its own lets you do whatever, but I got the feeling that the legacy version was only for backwards compatability with CIP-15 i.e. it shouldn't be there unless you're using weighted, and even then it might be optional?

Reply:
logically it looks like it's made for backwards compatibility, but it's cardano, if it turns out one day this thing can be used with legacy addresses, it can bring some trouble. @SebastienGllmt pls tell what you think when you have time